// fp-ts/es6/Const.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { fromEquals } from './Setoid';
import { phantom, toString } from './function';
export var URI = 'Const';
/**
 * @data
 * @constructor Const
 * @since 1.0.0
 */
var Const = /** @class */ (function () {
    function Const(value) {
        this.value = value;
    }
    Const.prototype.map = function (f) {
        return this;
    };
    Const.prototype.contramap = function (f) {
        return this;
    };
    Const.prototype.fold = function (f) {
        return f(this.value);
    };
    Const.prototype.inspect = function () {
        return this.toString();
    };
    Const.prototype.toString = function () {
        return "new Const(" + toString(this.value) + ")";
    };
    return Const;
}());
export { Const };
/**
 * @since 1.17.0
 */
export var getShow = function (S) {
    return {
        show: function (c) { return "new Const(" + S.show(c.value) + ")"; }
    };
};
/**
 * @since 1.0.0
 */
export var getSetoid = function (S) {
    return fromEquals(function (x, y) { return S.equals(x.value, y.value); });
};
var map = function (fa, f) {
    return fa.map(f);
};
var contramap = function (fa, f) {
    return fa.contramap(f);
};
var ap = function (S) { return function (fab, fa) {
    return new Const(S.concat(fab.value, fa.value));
}; };
/**
 * @since 1.0.0
 */
export var getApply = function (S) {
    return {
        URI: URI,
        _L: phantom,
        map: map,
        ap: ap(S)
    };
};
var of = function (M) { return function (a) {
    return new Const(M.empty);
}; };
/**
 * @since 1.0.0
 */
export var getApplicative = function (M) {
    return __assign({}, getApply(M), { of: of(M) });
};
/**
 * @since 1.0.0
 */
export var const_ = {
    URI: URI,
    map: map,
    contramap: contramap
};
