// fp-ts/es6/IOEither.js
import { left as eitherLeft, right as eitherRight, toError, tryCatch2v as eitherTryCatch2v } from './Either';
import * as eitherT from './EitherT';
import { constant, constIdentity } from './function';
import { IO, io } from './IO';
export var URI = 'IOEither';
var T = eitherT.getEitherT2v(io);
var foldT = eitherT.fold(io);
/**
 * @since 1.6.0
 */
var IOEither = /** @class */ (function () {
    function IOEither(value) {
        this.value = value;
    }
    /**
     * Runs the inner io
     */
    IOEither.prototype.run = function () {
        return this.value.run();
    };
    IOEither.prototype.map = function (f) {
        return new IOEither(T.map(this.value, f));
    };
    IOEither.prototype.ap = function (fab) {
        return new IOEither(T.ap(fab.value, this.value));
    };
    /**
     * Flipped version of `ap`
     */
    IOEither.prototype.ap_ = function (fb) {
        return fb.ap(this);
    };
    /**
     * Combine two effectful actions, keeping only the result of the first
     */
    IOEither.prototype.applyFirst = function (fb) {
        return fb.ap(this.map(constant));
    };
    /**
     * Combine two effectful actions, keeping only the result of the second
     */
    IOEither.prototype.applySecond = function (fb) {
        return fb.ap(this.map(constIdentity));
    };
    IOEither.prototype.chain = function (f) {
        return new IOEither(T.chain(this.value, function (a) { return f(a).value; }));
    };
    IOEither.prototype.fold = function (left, right) {
        return foldT(left, right, this.value);
    };
    IOEither.prototype.mapLeft = function (f) {
        return new IOEither(this.value.map(function (e) { return e.mapLeft(f); }));
    };
    IOEither.prototype.orElse = function (f) {
        return new IOEither(this.value.chain(function (e) { return e.fold(function (l) { return f(l).value; }, function (a) { return T.of(a); }); }));
    };
    IOEither.prototype.alt = function (fy) {
        return this.orElse(function () { return fy; });
    };
    IOEither.prototype.bimap = function (f, g) {
        return new IOEither(this.value.map(function (e) { return e.bimap(f, g); }));
    };
    return IOEither;
}());
export { IOEither };
var map = function (fa, f) {
    return fa.map(f);
};
var of = function (a) {
    return new IOEither(T.of(a));
};
var ap = function (fab, fa) {
    return fa.ap(fab);
};
var chain = function (fa, f) {
    return fa.chain(f);
};
var alt = function (fx, fy) {
    return fx.alt(fy);
};
var bimap = function (fa, f, g) {
    return fa.bimap(f, g);
};
/**
 * @since 1.6.0
 */
export var right = function (fa) {
    return new IOEither(fa.map(eitherRight));
};
/**
 * @since 1.6.0
 */
export var left = function (fa) {
    return new IOEither(fa.map(eitherLeft));
};
/**
 * @since 1.6.0
 */
export var fromEither = function (fa) {
    return new IOEither(io.of(fa));
};
/**
 * @since 1.6.0
 */
export var fromLeft = function (l) {
    return fromEither(eitherLeft(l));
};
/**
 * Use `tryCatch2v` instead
 *
 * @since 1.6.0
 * @deprecated
 */
export var tryCatch = function (f, onerror) {
    if (onerror === void 0) { onerror = toError; }
    return tryCatch2v(f, onerror);
};
/**
 * @since 1.11.0
 */
export var tryCatch2v = function (f, onerror) {
    return new IOEither(new IO(function () { return eitherTryCatch2v(f, onerror); }));
};
var throwError = fromLeft;
/**
 * @since 1.6.0
 */
export var ioEither = {
    URI: URI,
    bimap: bimap,
    map: map,
    of: of,
    ap: ap,
    chain: chain,
    alt: alt,
    throwError: throwError,
    fromEither: fromEither,
    fromOption: function (o, e) { return (o.isNone() ? throwError(e) : of(o.value)); }
};
