// fp-ts/es6/Record.js
import { getDictionaryMonoid } from './Monoid';
import { none, some as optionSome } from './Option';
import { fromEquals } from './Setoid';
import { showString } from './Show';
var _hasOwnProperty = Object.prototype.hasOwnProperty;
function hasOwnProperty(k, d) {
    return _hasOwnProperty.call(d, k);
}
/**
 * @since 1.17.0
 */
export var getShow = function (S) {
    return {
        show: function (r) {
            var elements = collect(r, function (k, a) { return showString.show(k) + ": " + S.show(a); }).join(', ');
            return elements === '' ? '{}' : "{ " + elements + " }";
        }
    };
};
/**
 * Calculate the number of key/value pairs in a record
 *
 * @since 1.10.0
 */
export var size = function (d) {
    return Object.keys(d).length;
};
/**
 * Test whether a record is empty
 *
 * @since 1.10.0
 */
export var isEmpty = function (d) {
    return Object.keys(d).length === 0;
};
export function collect(d, f) {
    var out = [];
    var keys = Object.keys(d).sort();
    for (var _i = 0, keys_1 = keys; _i < keys_1.length; _i++) {
        var key = keys_1[_i];
        out.push(f(key, d[key]));
    }
    return out;
}
export function toArray(d) {
    return collect(d, function (k, a) { return [k, a]; });
}
export function toUnfoldable(unfoldable) {
    return function (d) {
        var arr = toArray(d);
        var len = arr.length;
        return unfoldable.unfoldr(0, function (b) { return (b < len ? optionSome([arr[b], b + 1]) : none); });
    };
}
export function insert(k, a, d) {
    if (d[k] === a) {
        return d;
    }
    var r = Object.assign({}, d);
    r[k] = a;
    return r;
}
export function remove(k, d) {
    if (!hasOwnProperty(k, d)) {
        return d;
    }
    var r = Object.assign({}, d);
    delete r[k];
    return r;
}
/**
 * Delete a key and value from a map, returning the value as well as the subsequent map
 *
 * @since 1.10.0
 */
export var pop = function (k, d) {
    var a = lookup(k, d);
    return a.isNone() ? none : optionSome([a.value, remove(k, d)]);
};
/**
 * Test whether one record contains all of the keys and values contained in another record
 *
 * @since 1.14.0
 */
export var isSubrecord = function (S) { return function (d1, d2) {
    for (var k in d1) {
        if (!hasOwnProperty(k, d2) || !S.equals(d1[k], d2[k])) {
            return false;
        }
    }
    return true;
}; };
/**
 * Use `isSubrecord` instead
 * @since 1.10.0
 * @deprecated
 */
export var isSubdictionary = isSubrecord;
export function getSetoid(S) {
    var isSubrecordS = isSubrecord(S);
    return fromEquals(function (x, y) { return isSubrecordS(x, y) && isSubrecordS(y, x); });
}
export function getMonoid(S) {
    // tslint:disable-next-line: deprecation
    return getDictionaryMonoid(S);
}
/**
 * Lookup the value for a key in a record
 * @since 1.10.0
 */
export var lookup = function (key, fa) {
    return hasOwnProperty(key, fa) ? optionSome(fa[key]) : none;
};
export function filter(fa, p) {
    return filterWithKey(fa, function (_, a) { return p(a); });
}
/**
 * @since 1.10.0
 */
export var empty = {};
export function mapWithKey(fa, f) {
    var r = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_2 = keys; _i < keys_2.length; _i++) {
        var key = keys_2[_i];
        r[key] = f(key, fa[key]);
    }
    return r;
}
export function map(fa, f) {
    return mapWithKey(fa, function (_, a) { return f(a); });
}
/**
 * Reduce object by iterating over it's values.
 *
 * @since 1.10.0
 *
 * @example
 * import { reduce } from 'fp-ts/lib/Record'
 *
 * const joinAllVals = (ob: {[k: string]: string}) => reduce(ob, '', (acc, val) => acc + val)
 *
 * assert.deepStrictEqual(joinAllVals({a: 'foo', b: 'bar'}), 'foobar')
 */
export var reduce = function (fa, b, f) {
    return reduceWithKey(fa, b, function (_, b, a) { return f(b, a); });
};
/**
 * @since 1.10.0
 */
export var foldMap = function (M) {
    var foldMapWithKeyM = foldMapWithKey(M);
    return function (fa, f) { return foldMapWithKeyM(fa, function (_, a) { return f(a); }); };
};
/**
 * @since 1.10.0
 */
export var foldr = function (fa, b, f) {
    return foldrWithKey(fa, b, function (_, a, b) { return f(a, b); });
};
export function reduceWithKey(fa, b, f) {
    var out = b;
    var keys = Object.keys(fa).sort();
    var len = keys.length;
    for (var i = 0; i < len; i++) {
        var k = keys[i];
        out = f(k, out, fa[k]);
    }
    return out;
}
/**
 * @since 1.12.0
 */
export var foldMapWithKey = function (M) { return function (fa, f) {
    var out = M.empty;
    var keys = Object.keys(fa).sort();
    var len = keys.length;
    for (var i = 0; i < len; i++) {
        var k = keys[i];
        out = M.concat(out, f(k, fa[k]));
    }
    return out;
}; };
export function foldrWithKey(fa, b, f) {
    var out = b;
    var keys = Object.keys(fa).sort();
    var len = keys.length;
    for (var i = len - 1; i >= 0; i--) {
        var k = keys[i];
        out = f(k, fa[k], out);
    }
    return out;
}
/**
 * Create a record with one key/value pair
 *
 * @since 1.10.0
 */
export var singleton = function (k, a) {
    var _a;
    return _a = {}, _a[k] = a, _a;
};
export function traverseWithKey(F) {
    return function (ta, f) {
        var keys = Object.keys(ta);
        if (keys.length === 0) {
            return F.of(empty);
        }
        var fr = F.of({});
        var _loop_1 = function (key) {
            fr = F.ap(F.map(fr, function (r) { return function (b) {
                r[key] = b;
                return r;
            }; }), f(key, ta[key]));
        };
        for (var _i = 0, keys_3 = keys; _i < keys_3.length; _i++) {
            var key = keys_3[_i];
            _loop_1(key);
        }
        return fr;
    };
}
export function traverse(F) {
    var traverseWithKeyF = traverseWithKey(F);
    return function (ta, f) { return traverseWithKeyF(ta, function (_, a) { return f(a); }); };
}
export function sequence(F) {
    var traverseWithKeyF = traverseWithKey(F);
    return function (ta) { return traverseWithKeyF(ta, function (_, a) { return a; }); };
}
/**
 * @since 1.10.0
 */
export var compact = function (fa) {
    var r = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_4 = keys; _i < keys_4.length; _i++) {
        var key = keys_4[_i];
        var optionA = fa[key];
        if (optionA.isSome()) {
            r[key] = optionA.value;
        }
    }
    return r;
};
/**
 * @since 1.10.0
 */
export var partitionMap = function (fa, f) {
    return partitionMapWithKey(fa, function (_, a) { return f(a); });
};
/**
 * @since 1.10.0
 */
export var partition = function (fa, p) {
    return partitionWithKey(fa, function (_, a) { return p(a); });
};
/**
 * @since 1.10.0
 */
export var separate = function (fa) {
    var left = {};
    var right = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_5 = keys; _i < keys_5.length; _i++) {
        var key = keys_5[_i];
        var e = fa[key];
        if (e.isLeft()) {
            left[key] = e.value;
        }
        else {
            right[key] = e.value;
        }
    }
    return {
        left: left,
        right: right
    };
};
export function wither(F) {
    var traverseF = traverse(F);
    return function (wa, f) { return F.map(traverseF(wa, f), compact); };
}
export function wilt(F) {
    var traverseF = traverse(F);
    return function (wa, f) { return F.map(traverseF(wa, f), separate); };
}
/**
 * @since 1.10.0
 */
export var filterMap = function (fa, f) {
    return filterMapWithKey(fa, function (_, a) { return f(a); });
};
export function partitionMapWithKey(fa, f) {
    var left = {};
    var right = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_6 = keys; _i < keys_6.length; _i++) {
        var key = keys_6[_i];
        var e = f(key, fa[key]);
        if (e.isLeft()) {
            left[key] = e.value;
        }
        else {
            right[key] = e.value;
        }
    }
    return {
        left: left,
        right: right
    };
}
export function partitionWithKey(fa, p) {
    var left = {};
    var right = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_7 = keys; _i < keys_7.length; _i++) {
        var key = keys_7[_i];
        var a = fa[key];
        if (p(key, a)) {
            right[key] = a;
        }
        else {
            left[key] = a;
        }
    }
    return {
        left: left,
        right: right
    };
}
export function filterMapWithKey(fa, f) {
    var r = {};
    var keys = Object.keys(fa);
    for (var _i = 0, keys_8 = keys; _i < keys_8.length; _i++) {
        var key = keys_8[_i];
        var optionB = f(key, fa[key]);
        if (optionB.isSome()) {
            r[key] = optionB.value;
        }
    }
    return r;
}
export function filterWithKey(fa, p) {
    var r = {};
    var changed = false;
    for (var key in fa) {
        if (hasOwnProperty(key, fa)) {
            var a = fa[key];
            if (p(key, a)) {
                r[key] = a;
            }
            else {
                changed = true;
            }
        }
    }
    return changed ? r : fa;
}
export function fromFoldable(
// tslint:disable-next-line: deprecation
F) {
    return function (ta, f) {
        return F.reduce(ta, {}, function (b, _a) {
            var k = _a[0], a = _a[1];
            b[k] = hasOwnProperty(k, b) ? f(b[k], a) : a;
            return b;
        });
    };
}
export function fromFoldableMap(M, 
// tslint:disable-next-line: deprecation
F) {
    return function (ta, f) {
        return F.reduce(ta, {}, function (r, a) {
            var _a = f(a), k = _a[0], b = _a[1];
            r[k] = hasOwnProperty(k, r) ? M.concat(r[k], b) : b;
            return r;
        });
    };
}
/**
 * @since 1.14.0
 */
export function every(fa, predicate) {
    for (var k in fa) {
        if (!predicate(fa[k])) {
            return false;
        }
    }
    return true;
}
/**
 * @since 1.14.0
 */
export function some(fa, predicate) {
    for (var k in fa) {
        if (predicate(fa[k])) {
            return true;
        }
    }
    return false;
}
/**
 * @since 1.14.0
 */
export function elem(S) {
    return function (a, fa) { return some(fa, function (x) { return S.equals(x, a); }); };
}
export function partitionMapWithIndex(fa, f) {
    return partitionMapWithKey(fa, f);
}
export function partitionWithIndex(fa, p) {
    return partitionWithKey(fa, p);
}
export function filterMapWithIndex(fa, f) {
    return filterMapWithKey(fa, f);
}
export function filterWithIndex(fa, p) {
    return filterWithKey(fa, p);
}
