// fp-ts/es6/Writer.js
import { phantom } from './function';
export var URI = 'Writer';
/**
 * @since 1.0.0
 */
var Writer = /** @class */ (function () {
    function Writer(run) {
        this.run = run;
    }
    Writer.prototype.eval = function () {
        return this.run()[0];
    };
    Writer.prototype.exec = function () {
        return this.run()[1];
    };
    Writer.prototype.map = function (f) {
        var _this = this;
        return new Writer(function () {
            var _a = _this.run(), a = _a[0], w = _a[1];
            return [f(a), w];
        });
    };
    return Writer;
}());
export { Writer };
var map = function (fa, f) {
    return fa.map(f);
};
var of = function (M) { return function (a) {
    return new Writer(function () { return [a, M.empty]; });
}; };
var ap = function (S) { return function (fab, fa) {
    return new Writer(function () {
        var _a = fab.run(), f = _a[0], w1 = _a[1];
        var _b = fa.run(), a = _b[0], w2 = _b[1];
        return [f(a), S.concat(w1, w2)];
    });
}; };
var chain = function (S) { return function (fa, f) {
    return new Writer(function () {
        var _a = fa.run(), a = _a[0], w1 = _a[1];
        var _b = f(a).run(), b = _b[0], w2 = _b[1];
        return [b, S.concat(w1, w2)];
    });
}; };
/**
 * Appends a value to the accumulator
 *
 * @since 1.0.0
 */
export var tell = function (w) {
    return new Writer(function () { return [undefined, w]; });
};
/**
 * Modifies the result to include the changes to the accumulator
 *
 * @since 1.3.0
 */
export var listen = function (fa) {
    return new Writer(function () {
        var _a = fa.run(), a = _a[0], w = _a[1];
        return [[a, w], w];
    });
};
/**
 * Applies the returned function to the accumulator
 *
 * @since 1.3.0
 */
export var pass = function (fa) {
    return new Writer(function () {
        var _a = fa.run(), _b = _a[0], a = _b[0], f = _b[1], w = _a[1];
        return [a, f(w)];
    });
};
/**
 * Projects a value from modifications made to the accumulator during an action
 *
 * @since 1.3.0
 */
export var listens = function (fa, f) {
    return new Writer(function () {
        var _a = fa.run(), a = _a[0], w = _a[1];
        return [[a, f(w)], w];
    });
};
/**
 * Modify the final accumulator value by applying a function
 *
 * @since 1.3.0
 */
export var censor = function (fa, f) {
    return new Writer(function () {
        var _a = fa.run(), a = _a[0], w = _a[1];
        return [a, f(w)];
    });
};
/**
 *
 * @since 1.0.0
 */
export var getMonad = function (M) {
    return {
        URI: URI,
        _L: phantom,
        map: map,
        of: of(M),
        ap: ap(M),
        chain: chain(M)
    };
};
/**
 * @since 1.0.0
 */
export var writer = {
    URI: URI,
    map: map
};
